src/networking/systems.rs

```diff
@@ -0,0 +1,240 @@
+use std::{collections::HashMap};
+
+use bevy::prelude::{Commands, Entity, EventReader, Events, ResMut};
+use quinn::{crypto::rustls::TlsSession, generic::{RecvStream, SendStream}};
+use tokio::{stream::StreamExt, sync::mpsc::UnboundedReceiver, sync::mpsc::UnboundedSender, sync::mpsc::unbounded_channel};
+use tracing::{error, info};
+
+use super::{components::Connection, events::ReceiveEvent, events::SendEvent, id::ConnectionId, id::StreamId, packets::Packet, streams::BoundedPlanetRecvStream, streams::BoundedPlanetSendStream};
+
+/// Internal state of the network session system
+pub struct SessionEventListenerState {
+    /// Map from ConnectionID => StreamId => StreamSender
+    pub stream_senders: HashMap<ConnectionId, HashMap<StreamId, UnboundedSender<Packet>>>,
+
+    /// MPSC which receives all events from the network
+    pub event_receiver: UnboundedReceiver<ReceiveEvent>,
+
+    /// Event reader used for pulling send events and publishing them to the network
+    pub send_event_reader: EventReader<SendEvent>,
+}
+
+/// ECS resources containing a map of active network connections
+#[derive(Default)]
+pub struct NetworkConnections {
+    pub connections: HashMap<ConnectionId, Entity>
+}
+
+/// Consume events from ECS and publish them to the correct MPSC to send it over the network
+pub fn receive_net_events(
+    mut commands: Commands,
+    mut session: ResMut<SessionEventListenerState>,
+    mut entities: ResMut<NetworkConnections>,
+    mut net_events: ResMut<Events<ReceiveEvent>>
+) {
+    // Break up `session` in a way that Rust is happy with
+    let session: &mut SessionEventListenerState = &mut session;
+    let SessionEventListenerState { event_receiver, .. } = session;
+
+    // Pull network events from MPSC and publish them
+    while let Ok(event) = event_receiver.try_recv() {
```

Comment:
Is it possible this could get stuck in an infinite loop...?

Reply:
Only if there are an infinite number of events.